File: java/com/joshmanisdabomb/lcc/block/ClassicSaplingBlock.java
package com.joshmanisdabomb.lcc.block;

import com.joshmanisdabomb.lcc.registry.LCCBlocks;
import com.joshmanisdabomb.lcc.registry.LCCFeatures;
import net.minecraft.block.Block;
import net.minecraft.block.BlockState;
import net.minecraft.client.renderer.RenderType;
import net.minecraft.state.IntegerProperty;
import net.minecraft.state.StateContainer;
import net.minecraft.state.properties.BlockStateProperties;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import net.minecraft.world.gen.feature.Feature;
import net.minecraft.world.server.ServerWorld;

import java.util.Random;

public class ClassicSaplingBlock extends FunctionalSaplingBlock implements IPottableBlock, LCCBlockHelper {

    public static final IntegerProperty STAGE = BlockStateProperties.STAGE_0_1;
    public static final IntegerProperty AGE = BlockStateProperties.AGE_0_7;

    public ClassicSaplingBlock(Properties properties) {
        super((random, b) -> Feature.NORMAL_TREE.withConfiguration(LCCFeatures.CLASSIC_TREE_CONFIG), () -> LCCBlocks.potted_classic_sapling.getDefaultState(), properties);
        this.setDefaultState(this.stateContainer.getBaseState().with(STAGE, 0).with(AGE, 0));
    }

    @Override
    protected void fillStateContainer(StateContainer.Builder<Block, BlockState> builder) { builder.add(STAGE, AGE); }

    @Override
    public void tick(BlockState state, ServerWorld world, BlockPos pos, Random random) {
        if (!world.isAreaLoaded(pos, 1)) return;
        if (world.getLight(pos.up()) >= 9 && random.nextInt(5) == 0) {
            if (state.get(AGE) < 7) {
                world.setBlockState(pos, state.cycle(AGE), 3);
            } else if (state.get(STAGE) < 1) {
                world.setBlockState(pos, state.with(AGE, 0).cycle(STAGE), 3);
            } else {
                this.grow(world, random, pos, state);
            }
        }

    }

    @Override
    public void grow(ServerWorld world, Random rand, BlockPos pos, BlockState state) {
        this.func_226942_a_(world, pos, state.with(STAGE, 1), rand);
    }

    @Override
    public boolean canUseBonemeal(World worldIn, Random rand, BlockPos pos, BlockState state) {
        return true;
    }

    @Override
    public RenderType getRenderLayer() {
        return RenderType.getCutout();
    }

}

File: lcc-content/src/main/kotlin/com/joshmanisdabomb/lcc/block/ClassicSaplingBlock.kt
package com.joshmanisdabomb.lcc.block

import com.joshmanisdabomb.lcc.directory.LCCConfiguredFeatures
import net.minecraft.block.Block
import net.minecraft.block.BlockState
import net.minecraft.block.SaplingBlock
import net.minecraft.block.sapling.SaplingGenerator
import net.minecraft.server.world.ServerWorld
import net.minecraft.state.StateManager
import net.minecraft.state.property.Properties.AGE_7
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import java.util.*

class ClassicSaplingBlock(settings: Settings) : SaplingBlock(ClassicSaplingBlock, settings) {

    init {
        defaultState = this.stateManager.defaultState.with(STAGE, 0).with(AGE_7, 0)
    }

    override fun appendProperties(builder: StateManager.Builder<Block, BlockState>) = builder.add(STAGE, AGE_7).let {}

    override fun randomTick(state: BlockState, world: ServerWorld, pos: BlockPos, random: Random) {
        if (world.getLightLevel(pos.up()) >= 9 && random.nextInt(5) == 0) {
            if (state[AGE_7] < 7) {
                world.setBlockState(pos, state.cycle(AGE_7), 3)
            } else if (state[STAGE] < 1) {
                world.setBlockState(pos, state.with(AGE_7, 0).cycle(STAGE), 3)
            } else {
                grow(world, random, pos, state)
            }
        }
    }

    override fun canGrow(world: World, random: Random, pos: BlockPos, state: BlockState) = true

    override fun grow(world: ServerWorld, random: Random, pos: BlockPos, state: BlockState) {
        generate(world, pos, state.with(STAGE, 1), random)
    }

    companion object : SaplingGenerator() {

        override fun getTreeFeature(random: Random, bl: Boolean) = LCCConfiguredFeatures.classic_tree

    }

}